### PR TITLE
Optimize keys

### PR DESCRIPTION
**Pull request by angshuman**
- Optimize keys to reduce duplication
- Only write O key if O is Id
- Fix broken unit tests

### CODE REVIEW

File: Hexastore.Rocks/TripleExtensions.cs [hunk 1]
```diff
@@ -9,305 +8,149 @@ namespace Hexastore.Rocks
...
+                    break;
+                default:
+                    throw new InvalidOperationException("unmatched key. panic.");
+            }
 
-        public static byte[] GetOIsIdTypeBytes(this Triple triple)
+            return new Triple(s, p, new TripleObject(o, isId, type, index));
+        }
+
+        public static byte[] ToBytes(this Triple triple)
```
Comment (jobeland) [inline Hexastore.Rocks/TripleExtensions.cs]: Do we still need this? Don't see where we are serializing the entire triple to bytes anymore

File: Hexastore.Rocks/TripleExtensions.cs [hunk 2]
```diff
@@ -9,305 +8,149 @@ namespace Hexastore.Rocks
...
+            var spio = new[] { sBytes, pBytes, indexBytes, idBytes, oTypeBytes, oBytes };
 
             var index = 0;
-            foreach (var item in oIdType)
-            {
+            foreach (var item in spio) {
                 var lenBytes = BitConverter.GetBytes(item.Length);
-                Buffer.BlockCopy(lenBytes, 0, destination, index, BYTES_PER_INT);
-                index += BYTES_PER_INT;
+                Buffer.BlockCopy(lenBytes, 0, destination, index, 4);
```
Comment (jobeland) [inline Hexastore.Rocks/TripleExtensions.cs]: Nit: replace all the 4s with a const

File: Hexastore.Rocks/TripleObjectExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            // index is not converted
+            return KeyConfig.ConcatBytes(isIdBytes, typeBytes, oBytes);
+        }
+
+        public static TripleObject ToTripleObject(this byte[] input)
+        {
+            var isID = input[0] == KeyConfig.ByteTrue[0] ? true : false;
+            var typeSpan = new ReadOnlySpan<byte>(input, 1, 4);
+            var type = MemoryMarshal.Read<int>(typeSpan);
+            var valueLength = input.Length - 5;
```
Comment (jobeland) [inline Hexastore.Rocks/TripleObjectExtensions.cs:26]: Nit: I think these numbers represent the int byte size offset? Helpful for reading if they are named const and/or vars

File: Hexastore.Rocks/TripleExtensions.cs (hunk 1 above)
Comment (angshuman) [reply Hexastore.Rocks/TripleExtensions.cs]: Correct. Will remove.

File: Hexastore.Rocks/TripleExtensions.cs (hunk 2 above)
Comment (angshuman) [reply Hexastore.Rocks/TripleExtensions.cs]: I think its a bit too much. size of an int is always 4 and it makes it harder to read for me not knowing that.

File: Hexastore.Rocks/TripleObjectExtensions.cs (hunk 3 above)
Comment (angshuman) [reply Hexastore.Rocks/TripleObjectExtensions.cs:26]: I think the names makes it harder to read for me. Will keep 4 and 1 - those actually mean int or byte. Just read it like that.

**Pull request merged**